### ISSUE TITLE
Unable to detect multiple parameter annotations with different RetentionPolicies.

### ISSUE THREAD

**Issue by T101N** [classgraph/classgraph#335]
The issue is that parameter annotations with `RetentionPolicy.RUNTIME` doesn't work when the parameter also has another annotation with `RetentionPolicy.CLASS`.

Tested with master branch - version 4.8.22-SNAPSHOT (352eb88)

```
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.PARAMETER)
public @interface AnnoRuntime { }

@Retention(RetentionPolicy.CLASS)
@Target(ElementType.PARAMETER)
public @interface AnnoClass { }

public void function(@AnnoRuntime @AnnoClass int arg) { }
```

**Comment by lukehutch** [comment]
Thanks for the bug report, and especially for providing a unit test. This is quite weird behavior. I don't understand how one annotation's meta-annotation can affect another annotation. Do you have a theory about why this might be the case? Is it possible the classfile itself is compiled differently in this case? (I can't imagine it would be, so this is probably a ClassGraph bug, but I don't understand the cause.)

I'm traveling today (about to head to the airport), but will take a look at the unit tests in a debugger later.

**Comment by lukehutch** [comment]
Actually I had a thought: can you try switching the order of the annotations to see if that changes the behavior: it's possible this is indeed a compiler issue -- specifically that the retention policy is applied to the parameter, not the annotations, and all annotations inherit the retention policy of the last annotation, or something).

Currently only runtime-visible and runtime-invisible annotations are being processed. I didn't know about the `CLASS` retention policy, so that may explain the loss of these annotations, in combination with the last-takes-all behavior suggested above.

**Comment by T101N** [comment]
> Actually I had a thought: can you try switching the order of the annotations to see if that changes the behavior? It's possible this is indeed a compiler issue -- specifically that the retention policy is applied to the parameter, not the annotations, and all annotations inherit the retention policy of the last annotation, or something.
> 
> Currently only runtime-visible and runtime-invisible annotations are being processed. I didn't know about the `CLASS` retention policy, so that may explain the loss of these annotations, in combination with the last-takes-all behavior suggested above.

Sure, I'll supply a test with the annotations switched around. I think it's more to do with the construction of the AnnotationInfoList when scanning. I found this problem when playing around in Kotlin, where all parameters are annotated to @NotNull (RetentionPolicy.CLASS) by default. At first, I thought Kotlin was the issue, but when I checked its compiled class files, the annotation with the RUNTIME retention was still there. That's when I decided to test it in plain Java and encountered the same problem so I thought it's a bug with ClassGraph and wrote some quick tests for it.

**Comment by lukehutch** [comment]
Interesting.. No filtering is applied while an `AnnotationInfoList` is being constructed though, and there should be no state carried over from annotation to annotation. The parameter annotation info is read here:

https://github.com/classgraph/classgraph/blob/master/src/main/java/io/github/classgraph/Classfile.java#L1259

The `AnnotationInfoList` is built here:

https://github.com/classgraph/classgraph/blob/master/src/main/java/io/github/classgraph/MethodParameterInfo.java#L175

It's all pretty straightforward.. 

There have been Kotlin issues in the past, see:

https://github.com/classgraph/classgraph/blob/master/src/main/java/io/github/classgraph/MethodInfo.java#L375

**Comment by T101N** [comment]
Just tested it with the changed annotation ordering. It's still unable to detect the annotation with runtime retention.

> It's all pretty straightforward..

Hah, not to me. I'm terrible at file io, but thanks for the leads.

**Comment by lukehutch** [comment]
> Hah, not to me. I'm terrible at file io, but thanks for the leads.

Oh, I was only referring to the part that builds the `AnnotationInfoList` for each parameter. The classfile parsing code is nightmarishly complicated :-D

**Comment by lukehutch** [comment]
Weirdly the test passes fine in Eclipse, but it fails using `mvn test`...

**Comment by lukehutch** [comment]
Weirdly the test passes fine in Eclipse (compiled with ecj), but it fails using `mvn test`. It also fails in Eclipse if I build the code using commandline Maven, then run the test in Eclipse without rebuilding. So this is compiler-dependent somehow...

**Comment by lukehutch** [comment]
Even more weirdly, I managed to get `mvn test` to succeed once, then fail next time it was run. Nondeterminism is not good! Taking off on a flight now, will look more later.

**Comment by T101N** [comment]
That is strange indeed. Just tried it on Eclipse and got it to pass (right-click and run test), but now it fails after closing and reopening the project.

**Comment by lukehutch** [comment]
@T101N I found the source of the problem. In the classfile format, `RuntimeVisibleParameterAnnotations` and `RuntimeInvisibleParameterAnnotations` are listed in separate method attributes, and both of them can be present with separate arrays of annotations for the method. I was overwriting the method parameter annotations for the second attribute, rather than extending the arrays of annotations already read for the first attribute in order to merge together both annotation arrays into a single array of annotations for each method parameter.

Eclipse seems to only reliably fail if commandline Maven (commandline `javac`) last built the project, not `ecj`. However I saw some nondeterminacy between different runs on the commandline. Apparently `javac` uses a `Set` for storing annotations by retention type, and uses a nondeterministic hash function, probably based on the class reference (so the order in which classes are loaded can actually change the order that attributes are written to the classfile).

Released (and should be fixed) in 4.8.22. Thank you for the comprehensive testcase code!
